lock-api/src/main/java/com/palantir/lock/watch/LockEvent.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import java.util.Set;
+import java.util.function.Function;
+
+import org.immutables.value.Value;
+
+import com.palantir.lock.LockDescriptor;
+import com.palantir.lock.v2.LockToken;
+
+@Value.Immutable
+@Value.Style(visibility = Value.Style.ImplementationVisibility.PACKAGE)
+public abstract class LockEvent implements LockWatchEvent {
+    public abstract LockToken lockToken();
+    public abstract Set<LockDescriptor> lockDescriptors();
+
+    @Override
+    public void accept(LockWatchEventVisitor visitor) {
+        visitor.visit(this);
+    }
+
+    public static Function<Long, LockWatchEvent> fromSeq(LockToken lockToken, Set<LockDescriptor> lockDescriptors) {
```

Comment:
Does it make more sense to have a parallel `Builder` hierarchy for `LockWatchEvents`?

Reply:
The previous comment was a bit short and not very informative. We can use the fact that generated code with `@Value.Immutable` will generate a `ImmutableLockEvent.Builder` which would implement a `LockEvent.Builder` so instead of `fromSeq` we can have a static `builder` method. This would at least help us to remove `Function<Long, LockWatchEvent>` for something more intuitive.

Reply:
Maybe, but I don't really want to expose a generic builder. Maybe if we made it a staged builder